src/Compilers/Core/MSBuildTask/Csi.cs

```diff
@@ -0,0 +1,252 @@
+            }
+
+            get
+            {
+                return _store.GetOrDefault(nameof(CheckForOverflowUnderflow), false);
+            }
+        }
+
+        public ITaskItem[] UsingDirectives
+        {
+            set
+            {
+                _store[nameof(UsingDirectives)] = value;
+            }
+
+            get
+            {
+                return _store[nameof(UsingDirectives)] as ITaskItem[];
+            }
+        }
+        #endregion
+
+        #region Tool Members
+        /// <summary>
+        /// Return the name of the tool to execute.
+        /// </summary>
+        protected override string ToolName
+        {
+            get
+            {
+                return "Csi.exe";
+            }
+        }
+
+        /// <summary>
+        /// Return the path to the tool to execute.
+        /// </summary>
+        protected override string GenerateFullPathToTool()
+        {
+            string pathToTool = ToolLocationHelper.GetPathToBuildToolsFile(ToolName, ToolLocationHelper.CurrentToolsVersion);
+
+            if (pathToTool == null)
+            {
+                pathToTool = ToolLocationHelper.GetPathToDotNetFrameworkFile(ToolName, TargetDotNetFrameworkVersion.VersionLatest);
+
+                if (pathToTool == null)
+                {
+                    Log.LogErrorWithCodeFromResources("General_FrameworksFileNotFound", ToolName, ToolLocationHelper.GetDotNetFrameworkVersionFolderPrefix(TargetDotNetFrameworkVersion.VersionLatest));
+                }
+            }
+
+            return pathToTool;
+        }
+        #endregion
+
+        #region Interactive Compiler Members
+        protected override void AddResponseFileCommands(CommandLineBuilderExtension commandLine)
+        {
+            commandLine.AppendPlusOrMinusSwitch("/checked", _store, nameof(CheckForOverflowUnderflow));
+            commandLine.AppendSwitchUnquotedIfNotNull("/define:", GetDefineConstantsSwitch(DefineConstants));
+
+            AddReferencesToCommandLine(commandLine);
+            AddReferencePathsToCommandLine(commandLine);
+            AddUsingDirectivesToCommandLine(commandLine);
+
+            base.AddResponseFileCommands(commandLine);
+
+            // It's a good idea for the response file to be the very last switch passed, just 
+            // from a predictability perspective.  It also solves the problem that a dogfooder
+            // ran into, which is described in an email thread attached to bug VSWhidbey 146883.
+            // See also bugs 177762 and 118307 for additional bugs related to response file position.
+            if (ResponseFiles != null)
+            {
+                foreach (var response in ResponseFiles)
+                {
+                    commandLine.AppendSwitchIfNotNull("@", response.ItemSpec);
+                }
+            }
+        }
+        #endregion
+
+        private void AddReferencesToCommandLine(CommandLineBuilderExtension commandLine)
```

Comment:
Possibly. Csc has some extra lines for the EmbedInteropTypes and Link arguments, so that would need a switch.

It will probably be worth revisiting this task after @tmat finishes adding the rest of the command line argument support to csi.